src/sql/src/plan/query.rs

```diff
@@ -476,27 +478,93 @@ pub fn plan_insert_query(
     ))
 }
 
+/// Determines the mapping between some external data and a Materialize relation.
+///
+/// Returns the following:
+/// * [`CatalogItemId`] for the destination table.
+/// * [`RelationDesc`] representing the shape of the __input__ data we are copying from.
+/// * The [`ColumnIndex`]es that the source data maps to. TODO(cf2): We don't need this mapping
+///   since we now return a [`MapFilterProject`].
+/// * [`MapFilterProject`] which will map and project the input data to match the shape of the
+///   destination table.
+///
 pub fn plan_copy_item(
     scx: &StatementContext,
     item_name: ResolvedItemName,
     columns: Vec<Ident>,
-) -> Result<(CatalogItemId, RelationDesc, Vec<usize>), PlanError> {
+) -> Result<
+    (
+        CatalogItemId,
+        RelationDesc,
+        Vec<ColumnIndex>,
+        Option<MapFilterProject>,
+    ),
+    PlanError,
+> {
     let item = scx.get_item_by_resolved_name(&item_name)?;
+    let fullname = scx.catalog.resolve_full_name(item.name());
+    let table_desc = item.desc(&fullname)?.into_owned();
+    let mut ordering = Vec::with_capacity(columns.len());
 
-    let mut desc = item
-        .desc(&scx.catalog.resolve_full_name(item.name()))?
-        .into_owned();
+    // TODO(cf2): The logic here to create the `source_desc` and the MFP are a bit duplicated and
+    // should be simplified. The reason they are currently separate code paths is so we can roll
+    // out `COPY ... FROM <url>` without touching the current `COPY ... FROM ... STDIN` behavior.
 
-    let mut ordering = Vec::with_capacity(columns.len());
+    // If we're copying data into a table, then we generate an MFP.
+    let mfp = if let Some(table_defaults) = item.writable_table_details() {
```

Comment:
The comment says that we do something in case the item is a table but the code checks something more specific. I'm not super familiar with this but my initial doubt is whether there are tables for which writable_table_details returns None. Can this happen? Should we check if the item is a table and then assert that we get table_defaults inside?

Reply:
It can happen! For example objects created via `CREATE TABLE ... FROM SOURCE` will return `None` here. I updated the comment to explain some more of that context